s3parq/publish_parq.py

```diff
@@ -232,59 +388,85 @@ def log_size_estimate(num_bytes):
-            cluster_id: str
-            host: str 
-            port: str 
-            db_name: str
+        yield {'lower': lower, 'upper': upper}
+
+
+def publish(bucket: str, key: str, partitions: List[str], dataframe: pd.DataFrame, redshift_params: dict = None) -> List[str]:
+    """ Dataframe to S3 Parquet Publisher
+    This function handles the portion of work that will see a dataframe converted
+    to parquet and then published to the given S3 location.
+    It supports partitions and will preserve the datatypes on s3parq fetch.
+    It also has the option to automatically publish up to Redshift Spectrum for
+    the newly published parquet files.
+
+    Args:
+        bucket (str): S3 Bucket name
+        key (str): S3 key to lead to the desired dataset
+        partitions (List[str]): List of columns that should be partitioned on
+        dataframe (pd.DataFrame): Dataframe to be published
+        redshift_params (dict, Optional):
+            This dictionary should be provided in the following format in order
+            for data to be published to Spectrum. Leave out entirely to avoid
+            publishing to Spectrum.
+            The params should be formatted as follows:
+                schema_name (str): Name of the Spectrum schema to publish to
+                table_name (str): Name of the table to write the dataset as
+                iam_role (str): Role to take while writing data to Spectrum
+                region (str): AWS region for Spectrum
+                cluster_id (str): Spectrum cluster id
+                host (str): Redshift Spectrum host name
+                port (str): Redshift Spectrum port to use
+                db_name (str): Redshift Spectrum database name to use
+                ec2_user (str): If on ec2, the user that should be used
+
+    Returns:
+        A str list of all the newly published object keys
     """
     session_helper = None
 
     if redshift_params:
         if "index" in dataframe.columns:
-            raise ValueError("'index' is a reserved keyword in Redshift. Please remove or rename your DataFrame's 'index' column.")
+            raise ValueError(
+                "'index' is a reserved keyword in Redshift. Please remove or rename your DataFrame's 'index' column.")
 
-        logger.debug("Found redshift parameters. Checking validity of params...")
+        logger.debug(
+            "Found redshift parameters. Checking validity of params...")
         redshift_params = validate_redshift_params(redshift_params)
         logger.debug("Redshift parameters valid. Opening Session helper.")
         session_helper = SessionHelper(
-            region = redshift_params['region'],
-            cluster_id = redshift_params['cluster_id'],
-            host = redshift_params['host'],
-            port = redshift_params['port'],
-            db_name = redshift_params['db_name'],
-            ec2_user = redshift_params['ec2_user']
+            region=redshift_params['region'],
```

Comment:
you don't like my spaces? :(

Reply:
Full disclosure, I do prefer the spaces but pep8 does not :( (I auto-pep8-ed the files I touched)

Reply:
I see I see